llumnix/entrypoints/vllm/api_server.py

```diff
@@ -241,22 +241,30 @@ async def is_ready():
         # Launch the ray cluster for multi-node serving.
         launch_ray_cluster(args.ray_cluster_port)
 
+    # Connect to a ray cluster.
+    connect_to_ray_cluster(port=args.ray_cluster_port)
+
     # if gpu is not available, it means that this node is head pod without any llumnix components
     if is_gpu_available():
         # Launch the Llumnix componets on current node.
         server_id = random_uuid()
         node_id = ray.get_runtime_context().get_node_id()
-        engine_manager, instance_ids, llumlets, request_output_queue = init_llumnix_components(engine_manager_args, engine_args, node_id)
-        for idx, ins_id in enumerate(instance_ids):
-            instances[ins_id] = llumlets[idx]
-            instance_num_request[ins_id] = 0
-        log_requests = not args.disable_log_requests_server
-        # Start the api server after all the components of llumnix are ready.
-        print(f"Start Api Server on '{args.host}:{args.port}'")
-        uvicorn.run(app,
-                    host=args.host,
-                    port=args.port,
-                    log_level="debug",
-                    timeout_keep_alive=TIMEOUT_KEEP_ALIVE,
-                    ssl_keyfile=args.ssl_keyfile,
-                    ssl_certfile=args.ssl_certfile)
+        engine_manager, instance_ids, llumlets, request_output_queue, llumlets_avaliable = \
+            init_llumnix_components(engine_manager_args, engine_args, node_id)
+
+        if llumlets_avaliable:
+            for idx, ins_id in enumerate(instance_ids):
+                instances[ins_id] = llumlets[idx]
+                instance_num_request[ins_id] = 0
+            log_requests = not args.disable_log_requests_server
+            # Start the api server after all the components of llumnix are ready.
+            print(f"Start Api Server on '{args.host}:{args.port}'")
+            uvicorn.run(app,
+                        host=args.host,
+                        port=args.port,
+                        log_level="debug",
+                        timeout_keep_alive=TIMEOUT_KEEP_ALIVE,
+                        ssl_keyfile=args.ssl_keyfile,
+                        ssl_certfile=args.ssl_certfile)
+        else:
+            print("No llumlet avaliable, exit")
```

Comment:
Launch api server no matter if there is llumlet available.

Reply:
get